### PR TITLE
enhancement/rangeselector-update-redraw

### PR DESCRIPTION
**Pull request by pawellysy**
Implemented new flow for update for RangeSelector.

TODO:

- [x] Visual Diffs
- [x] dropdown update

### CODE REVIEW

File: ts/Stock/RangeSelector/RangeSelectorComposition.ts [hunk 1]
```diff
@@ -278,15 +278,17 @@ function onChartGetMargins(
 ): void {
     const rangeSelector = this.rangeSelector;
 
-    if (rangeSelector) {
+    if (rangeSelector && rangeSelector.options?.enabled) {
```
Comment (kamil-musialowski) [inline ts/Stock/RangeSelector/RangeSelectorComposition.ts:281]:
```suggestion
    if (rangeSelector?.options?.enabled) {
```

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 2]
```diff
@@ -1258,22 +1263,41 @@ class RangeSelector {
...
     ): void {
 
+        if (this.options.enabled === false) {
+            return;
+        }
+
         const chart = this.chart,
             chartOptions = chart.options,
             options =
-                chartOptions.rangeSelector as RangeSelectorOptions,
```
Comment (kamil-musialowski) [inline ts/Stock/RangeSelector/RangeSelector.ts:1264]: I think we need this spacing here :D

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 3]
```diff
@@ -1506,9 +1570,99 @@ class RangeSelector {
...
+            // Destroy additional buttons
+            if (newButtonsOptions.length < this.buttonOptions.length) {
+                for (
+                    let i = this.buttonOptions.length - 1;
+                    i >= newButtonsOptions.length;
+                    i--
+                ) {
+                    const btn = this.buttons.pop();
+                    btn!.destroy();
+                    this.dropdown!.options.remove(i + 1);
```
Comment (kamil-musialowski) [inline ts/Stock/RangeSelector/RangeSelector.ts:1625]: How about optional chain instead of force non-null?

**Review by kamil-musialowski — CHANGES_REQUESTED** [review]
Minor stuff, great work 👍

**Review by bm64 — APPROVED** [review]

**Review by pawelfus — CHANGES_REQUESTED** [review]
Visual changes are still failing.
As agreed, I'm okay with matching designs in styledMode and normal one, however I don't see related changes in CSS file

**Review by kamil-musialowski — APPROVED** [review]

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 4]
```diff
@@ -1506,9 +1569,99 @@ class RangeSelector {
...
+
+        if (!chart.styledMode) {
+            this.maxDateBox?.attr({
+                stroke: inputBoxBorderColor
+            });
+
+            this.minDateBox?.attr({
+                stroke: inputBoxBorderColor
+            });
+        }
```
Comment (pawelfus) [inline ts/Stock/RangeSelector/RangeSelector.ts:1593]: The height and stroke attribs shouldn't be part of `alignElements` method. Please, keep this method clean and short

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 5]
```diff
@@ -1506,9 +1569,99 @@ class RangeSelector {
...
+                    this.computeButtonRange(newButtonsOptions[i]);
+                }
+            }
+            this.buttonOptions = this.options.buttons ?? [];
+
+            if (defined(this.options.selected)) {
+                this.clickButton(this.options.selected, false);
+            }
+
+        }
```
Comment (pawelfus) [inline ts/Stock/RangeSelector/RangeSelector.ts:1662]: This whole block of code should be outside of this function.

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 6]
```diff
@@ -1612,9 +1765,9 @@ class RangeSelector {
 
                 // Skip animation
                 inputGroup.placed = chart.hasLoaded;
+                this.handleCollision(xOffsetForExportButton);
```
Comment (pawelfus) [inline ts/Stock/RangeSelector/RangeSelector.ts:1768]: Incorrect place. Buttons can be aligned to the right (eg rtl languages) and collision detection should include them no matter if inputs are enabled

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 7]
```diff
@@ -1585,7 +1738,7 @@ class RangeSelector {
 
             let xOffsetForExportButton = 0;
 
-            if (inputGroup) {
+            if (options.inputEnabled && inputGroup) {
```
Comment (pawelfus) [inline ts/Stock/RangeSelector/RangeSelector.ts:1741]: Q: Did something change I missed? `input.group` shouldn't exist if inputs are disabled, right?

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 8]
```diff
@@ -2069,9 +2223,20 @@ class RangeSelector {
     ): void {
         const chart = this.chart;
         merge(true, chart.options.rangeSelector, options);
+        merge(true, this.options, options);
```
Comment (pawelfus) [inline ts/Stock/RangeSelector/RangeSelector.ts:2226]: Do we need to merge it twice? Could something like this be enough?
```js
this.options = chart.options.rangeSelector = merge(chart.options.rangeSelector, options);
```

I assume those are references to the same object anyway, according to `init()` method

File: ts/Stock/RangeSelector/RangeSelector.ts [hunk 9]
```diff
@@ -2069,9 +2223,20 @@ class RangeSelector {
...
+        if (
+            this.options.selected &&
+            this.options.selected >= this.options.buttons!.length
+        ) {
+            this.options.selected = void 0;
+            chart.options.rangeSelector!.selected = void 0;
```
Comment (pawelfus) [inline ts/Stock/RangeSelector/RangeSelector.ts:2232]: With the change about `merge()` above, only `this.options.selected` could be changed. I see a potential bug with `selected = 0`, but that could be a misconfigured chart. Please make sure `update({ selected: 0, buttons : [] })` doesn't throw any errors

**Review by pawelfus — CHANGES_REQUESTED** [review]
Nice work! A few things to improve, see comments inline

**Review by DJTechnoo — APPROVED** [review]
Amazing work!

File: ts/Stock/RangeSelector/RangeSelector.ts (hunk 4 above)
Comment (pawellysy) [reply ts/Stock/RangeSelector/RangeSelector.ts:1593]: The stroke can be elsewhere, but the height may have an impact to the rangeSelector alignment, so I can't move it.

File: ts/Stock/RangeSelector/RangeSelector.ts (hunk 6 above)
Comment (pawellysy) [reply ts/Stock/RangeSelector/RangeSelector.ts:1675]: But this method is supposed to handle collision between the button group and input group only.

File: ts/Stock/RangeSelector/RangeSelector.ts (hunk 7 above)
Comment (pawellysy) [reply ts/Stock/RangeSelector/RangeSelector.ts:1648]: The `alignElements` method is fired multiple times. On the update it is fired 2 times. If you update the input to be disabled, in the first iteration the `options.inputEnabled`option if false, but `inputGroup`still exists. It gets deleted later. So I did it in sake performance.

**Review by pawelfus — COMMENTED** [review]
Looks good, awaiting benchmarks

**Review by pawelfus — CHANGES_REQUESTED** [review]
Could you check why other benchmarks run slower in this PR? Is this related to the benches itself, or did we break something by accident?

**Comment by pawellysy** [comment]
@pawelfus I've checked the benchmarks in the browser and the results are as follows:

### Navigator update:
- on v11.4.2 - 1765 - 1840
- on branch - 1450 - 1480
- on master -1465

### Stock tools
- on v11.4.2 - 1462 - 1505
- on branch - 1468 - 1491
- on master -1461 - 1500


### Stock update
- on v11.4.2 - 2500 - 2531
- on branch - 2450 - 2510
- on master -2350 - 2510

in my opinion, the differences between the results are neglectable in this example. Obviously, we are having some kind of memory leak problem, so maybe the jsdom env behaves differently than the native browser.

**Comment by pawellysy** [comment]
I've tested the PR against the current master using the same test as `update-navigator` and the results are as follows : 

### Sample size 1k
| Test | avg (ms) |
| --- | --- |
| update on local branch | 5991
| update on release | 5933

### Sample size 10k
| Test | avg (ms) |
| --- | --- |
| update on local branch | 3129
| update on release | 3252

### Sample size 100k
| Test | avg (ms) |
| --- | --- |
| update on local branch | 3847
| update on release | 3897



I think that the results are close, not like the ones shown in the Benchmark above.
In my opinion it is a proof that there is something wrong with our benchmark environment. @pawelfus

**Review by pawelfus — APPROVED** [review]
Approved. Thank you for checking this!

We have to fix/change/workaround the problem with JSDOM somewhere in the nearest future: [Ticket](https://app.asana.com/0/1205411033697899/1207617604993877/f)

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, great work!

The tests fail on CI though, do you know what causes that? Doesn't seem like a specific unit test it failing, but the suite fails.

**Comment by pawellysy** [comment]
@TorsteinHonsi seems like there is an issue with resize observer. I will investigate, what the issue is 
image

**Comment by pawellysy** [comment]
@TorsteinHonsi I merged the master branch, and now the previous issue disappeared. Now only failing tests are the ones, that are also failing on master.

**Review by TorsteinHonsi — APPROVED** [review]
🎉

**Pull request merged**